### ISSUE TITLE
 mMapView.zoomToBoundingBox is not calculating the required zoom level correctly

### ISSUE THREAD

**Issue by spyhunter99** [osmdroid/osmdroid#683]
I have a relatively tight bounds on NYC and for some reason, i keep ending up at zoom level 2. This may be an artifact of the floating point zoom level stuff, but i'm not sure at this point

**Comment by monsieurtanuki** [comment]
Could you give an example including the `BoundingBox.toString()`?

**Comment by spyhunter99** [comment]
good idea, smells like it needs a unit test

**Comment by spyhunter99** [comment]
![image](https://user-images.githubusercontent.com/1975321/29754645-62456c48-8b57-11e7-9987-5ea7a5c4ec14.png)

**Comment by spyhunter99** [comment]
the current bounding box is jacked up obviously. This is happening from a sample fragement in the method addOverlays which fires in onActivityCreated. The same code works ok if it happens later in the life cycle so i think it may be one of the things related to the first layout thing

**Comment by spyhunter99** [comment]
duplicate apparently
https://github.com/osmdroid/osmdroid/issues/236

**Comment by monsieurtanuki** [comment]
I think there's a combination of different potential bugs:
* I'm not convinced by the code of `MapView.zoomToBoundingBox` (for instance why should we use `mZoomLevel`?) and I'm going to refactor it partly in `TileSystem` and make it testable
* there's the notorious "I need the screen dimensions to compute this but I compute this before the MapView has dimensions" bug that explains the lousy `currentBox` values
* calls to `getController().zoomTo` and `getController().setZoom` cast as `int` the zoom computed as `double`, which obviously leads to a lack of accuracy in most cases

**Comment by spyhunter99** [comment]
I think it attempts good point, the current zoom level shouldn't matter for determining what the needed zoom level is

**Comment by monsieurtanuki** [comment]
I thought it would take an hour, it took a whole day...
In the process I dug up interesting things.

I fixed `MapView.zoomToBoundingBox`
Now the code is in `TileSystem.getBoundingBoxZoom`. In the previous code there was something fishy (presence of unnecessary `mZoomLevel`) and something obviously forgotten (we computed a zoom as a `double` but recasted it as `int`). And the code was not testable.
But after I started I realized that the code needed more than refactoring. In order to compute the zoom we need in order to display two longitudes, knowing the longitude span is enough. But for latitudes, as LatToY is not linear we cannot rely on the latitude span, we need the north and the south values.

Another effect of LatToY not being linear, if we center a map on a given GeoPoint, and if we display a `Polygon` which has the same center as the map, the `Polygon` won't necessarily be centered in the map on the Y-axis. It will look OK in city zoom levels, but if your map spans from N80 to S20 the `Polygon` will be displayed a little higher.

`BoundingBox.getCenter` doesn't take into account the notions of East and West: the resulting longitude is just half the sum of East and West. I created method `getCenterWithDateLine` in order to get the real center when the West is in Japan and the East is in Alaska (more or less the longitude of New Caledonia).

There's a side effect for `TileSystem.MinLatitude` (`-85.05112878`) and `TileSystem.MaxLatitude` (`-85.05112878`).
If you use the standard calculation `0.5 - log((1 + sin(lat)) / (1 - sin(lat))) / (4 * PI)` you find a slightly negative value for the max latitude (`-6.228684235054516E-12`) (expected value: 0+) and a value slightly above 1 for the min latitude (`1.0000000000062288`) (expected value : 1-).
If we use `-85.05112877` and `85.05112877` (as found elsewhere in the web) the results are in [0,1] (`3.1577035430885303E-10` and `0.9999999996842297`).
It shouldn't be an issue now as the results are then truncated into [0,1].
I discovered that when my unit tests failed. Maybe someone will find that interesting ;)
For instance if we multiply those results by the size of the map in pixels, maybe there will be rounding side effects.

I created `TileSystemTest` for `getBoundingBoxZoom` and each time the test is run, I run 2000 different checks with random values. Therefore I'm rather confident that the results are good.

I refreshed the demo "More Samples / Events / Zoom to Bounding Box": now it's fun to run and it's easy to check that the results look ok. Except for the MapView center vs. Polygon center offset I mentioned above, and there's nothing we can do about it.
